sources/osgUtil/IntersectFunctor.js

```diff
@@ -0,0 +1,152 @@
+'use strict';
+
+var vec3 = require('osg/glMatrix').vec3;
+var mat4 = require('osg/glMatrix').mat4;
+var PrimitiveIndexFunctor = require('osg/PrimitiveIndexFunctor');
+var intersectionEnums = require('osgUtil/intersectionEnums');
+
+// Settings are needed.
+var IntersectFunctor = function() {
+    this._hit = false;
+
+    // geom info
+    this._primitiveIndex = 0;
+    this._vertices = undefined;
+    this._geometry = undefined;
+
+    // intersection visitor (nodepath, etc...)
+    this._intersectionVisitor = undefined;
+
+    // intersector stuffs
+    this._intersector = undefined;
+    this._limitOneIntersection = false;
+    this._primitiveMask = intersectionEnums.ALL_PRIMITIVES;
+
+    this._Intersection = IntersectFunctor.Intersection;
+};
+
+// minimal "interface" for intersection type
+IntersectFunctor.Intersection = function() {
+    this._nodePath = undefined;
+    this._drawable = undefined;
+    this._matrix = undefined;
+    this._primitiveIndex = undefined;
+
+    this._ratio = 0.0;
+    this._backface = false;
+    this._localIntersectionPoint = vec3.create();
+};
+
+IntersectFunctor.prototype = {
+    reset: function() {
+        this._hit = false;
+        this._vertices = undefined;
+        this._primitiveIndex = 0;
+    },
+
+    setPrimitiveIndex: function(primitiveIndex) {
+        this._primitiveIndex = primitiveIndex;
+    },
+    setGeometry: function(geometry) {
+        this._geometry = geometry;
+        this._primitiveIndex = 0;
+        this._vertices = geometry.getAttributes().Vertex.getElements();
+    },
+    setVertices: function(vertices) {
+        this._vertices = vertices;
+    },
+
+    setIntersectionVisitor: function(intersectionVisitor) {
+        this._intersectionVisitor = intersectionVisitor;
+    },
+    setIntersector: function(intersector) {
+        this._intersector = intersector;
+        this._primitiveMask = intersector.getPrimitiveMask();
+
+        var limit = intersector.getIntersectionLimit();
+        this._limitOneIntersection =
+            limit === intersectionEnums.LIMIT_ONE_PER_DRAWABLE ||
+            limit === intersectionEnums.LIMIT_ONE;
+    },
+
+    setLimitOneIntersection: function(limitOneIntersection) {
+        this._limitOneIntersection = limitOneIntersection;
+    },
+
+    leave: function() {},
+
+    enter: function(/*bbox*/) {},
+
+    _createIntersection: function() {
+        var intersection = new this._Intersection();
```

Comment:
I think it could be clearer to have a void method  "createIntesection" here and override in derived classes.  Otw not sure about the naming, is a bit confusing.

Reply:
@cedricpinson